### PR TITLE
Update QueryUtils to use CollectorManager

### PR DESCRIPTION
**Pull request by harshavamsi**
### Description

Relates to https://github.com/apache/lucene/issues/12892

We introduce a new `SimpleCollectorWithLastReader` class to help keep context of LastReader per thread. This way calling `CollectorManager#reduce` can work with an executor. 

The diff makes it look big, but in reality we just wrap SimpleCollector inside a CollectorManager.

### CODE REVIEW

File: lucene/test-framework/src/java/org/apache/lucene/tests/search/QueryUtils.java [hunk 1]
```diff
@@ -345,175 +348,199 @@ public static void checkSkipTo(final Query q, final IndexSearcher s) throws IOEx
...
-                  for (int d = iterator.advance(lastDoc[0] + 1);
-                      d != DocIdSetIterator.NO_MORE_DOCS;
-                      d = iterator.nextDoc()) {
-                    if (liveDocs == null || liveDocs.get(d)) {
-                      more = true;
-                      break;
+                    @Override
+                    public LeafReader getLastReader() {
+                      return null;
                     }
```
Comment (msfroh) [inline lucene/test-framework/src/java/org/apache/lucene/tests/search/QueryUtils.java:516]: You'll need to return `lastReader`.

File: lucene/test-framework/src/java/org/apache/lucene/tests/search/QueryUtils.java [hunk 2]
```diff
@@ -540,117 +567,148 @@ public static void checkFirstSkipTo(final Query q, final IndexSearcher s) throws
...
 
-            lastReader[0] = context.reader();
-            lastDoc[0] = -1;
-          }
-        });
+              @Override
+              public List<LeafReader> reduce(Collection<SimpleCollectorWithLastReader> collectors) {
+                List<LeafReader> lastReaders = new ArrayList<>();
+                for (SimpleCollectorWithLastReader collector : collectors) {
+                  lastReaders.add(collector.getLastReader());
```
Comment (msfroh) [inline lucene/test-framework/src/java/org/apache/lucene/tests/search/QueryUtils.java:695]: You may want to check if the value returned from `getLastReader` is `null` before adding it to the array (and in the other reduce method).

**Comment by msfroh** [comment]
I had another thought while walking home. Both of the anonymous CollectorManagers are the same (but produce slightly different collectors). What if you split it out into an inner class that takes a `Supplier<SimpleCollectorWithLeafReader>` as a constructor arg?

That way the two checks can each have their anonymous implementations, but the anonymous nesting won't be as deep.

File: lucene/test-framework/src/java/org/apache/lucene/tests/search/QueryUtils.java [hunk 3]
```diff
@@ -345,175 +348,201 @@ public static void checkSkipTo(final Query q, final IndexSearcher s) throws IOEx
...
-                              + "  "
-                              + q.getClass().getName()
-                              + "\n\t Searcher="
-                              + s
-                              + "\n\t Order="
-                              + sbord
-                              + "\n\t Op="
-                              + (op == skip_op ? " skip()" : " next()"));
+                      lastReader = context.reader();
+                      assert readerContextArray.get(leafPtr).reader() == context.reader();
```
Comment (msfroh) [inline lucene/test-framework/src/java/org/apache/lucene/tests/search/QueryUtils.java:508]: This assert is failing. 

I assume it's because each slice gets a subset of segments, but the old logic assumed that the LeafReaders would be passed to the collector in the same order that they're in the top-level reader.

In a concurrent search world, I don't think this assertion adds value. As long as all the segments get collected, I don't think the order really matters.

Comment (javanna) [reply lucene/test-framework/src/java/org/apache/lucene/tests/search/QueryUtils.java:508]: I did not have time to review this in depth, but there is more to it than just the assertion. These collectors are rather hard to follow at first glance. Also, this is a case where we cannot resort to a single-threaded collector manager unfortunately, because the searcher is externally provided and that would require disabling concurrency in all those tests that call `QueryUtils` methods somehow. Calling search(Query, Collector) bypassed concurrency so far despite an executor is provided to the searcher, which is why we want to remove it.
